[DNM] [WIP] Small addition to string interpolation

beccadax opened the pull request:
This is just #18590 with an unconstrained `init(stringLiteral:)` which uses the `StringInterpolation` type and `init(stringInterpolation:)`. I'm going to take a baseline benchmark, then disable the built-in`DefaultStringInterpolation` default implementations and see how much overhead the unconstrained one adds. If we decide to keep anything from this PR, it'll be added to #18590.

beccadax committed:
Tweak TypeChecker::callWitness() signature

It doesn’t actually need its array to be mutable.

beccadax committed:
Character->String through a less circular path

beccadax committed:
Extract NSCoding unstable name diagnosis

We’ll want to add more code near this, so it’s threatening to get unweildy.

beccadax committed:
Add TapExpr

This is not used for anything yet, but it will contain the append* calls on the StringInterpolation instance.

beccadax committed:
Permit empty interpolations

As in, “foo\()baz”. These were previously hard syntax errors due to a quirk in Parser::parseList(); there’s no particular reason they should be.

beccadax committed:
Refactor getCalledValue()

To also support a not-yet-used getCalledDeclRef().

beccadax committed:
Rework ExpressibleByStringInterpolation

beccadax committed:
Rip out typeCheckExpressionShallow()

It is now dead code.

beccadax committed:
Tweak fix-it locations for throwing interpolators

beccadax committed:
Diagnose missing appendInterpolation methods

Since we can’t declare a formal appendInterpolation(…) requirement, we manually check that there is at least one method which:

* is as public as the type itself,
* returns Void or has an @discardableResult, and
* is not static.

If there isn’t, we emit an error on the conformance and a warning for each appendInterpolation method present, indicating why it is not suitable.

beccadax committed:
Warn about strange interpolations in Swift 4.2

beccadax committed:
Late-breaking string interpolation test fixes

beccadax committed:
Conform float types to TextOutputStreamable

This should improve interpolation performance, but I don’t know what impact the refactoring might have on .description, so we’ll see if we keep it.

beccadax committed:
Remove @_effects(readonly) helper functions

New string interpolation design can’t use @_effects(readonly), so these are no longer useful.

beccadax committed:
Apply feedback from @slavapestov

beccadax committed:
Simplify InterpolatedStringLiteralExpr::forEachSegment()

Removes the helper InterpolatedStringLiteralExpr::SegmentInfo struct. This also allows us to bring back the old ApplyExpr::getCalledValue() implementation.

beccadax committed:
Fix issues from rebasing

beccadax committed:
Update for DeclContext renaming

beccadax committed:
Fix method call on null object

beccadax committed:
Make both same-file extensions and default args work

Previously, the parser generated compound method names for appendInterpolation(…) methods because this helped when finding appendInterpolation methods declared in the same file. However, this turned out to prevent default arguments from working.

This commit returns it to adding base names only and instead explicitly calls loadAllMembers() on the StringInterpolation type. I’m not sure why types in other expressions don’t need this but types in these generated expressions do, but it’s much closer to the problem and doesn’t seem to have any ill effects.

beccadax committed:
[NFC] Improve debuggability of lookupDirect()

This change:

1. Adds accessors for the bit attached to NominalTypeDecl::LookupTable. These let you more easily search for and break on changes; more importantly, they give this bit a name and some documentation explaining what (I think) it means.

2. Includes the “ignoreNewExtensions” parameter in the debug output from NominalTypeDecl::lookupDirect().

3. Adds a MemberLookupTable::dump() method.

beccadax committed:
[NFC] Remove dead MemberLookupTable::addExtensionMembers()

These methods are never used and appear to be holdovers from an earlier implementation.

beccadax committed:
[AST] Update lazy LookupTable when extension added

Previously, nothing would guarantee that, if a nominal type `T` with lazy members had extensions adding members with name `foo`, and these members were already in `T`’s LookupTable, and a new extension to `T` added another member named `foo`, the new extension’s member would be added to the LookupTable.

This change makes it so that adding an extension clears the isLookupTablePopulated() flag, and so that when the flag is cleared on a type with lazy members, it updates all existing entries instead of just the ones present in the type itself.

Finally, it removes a workaround in string interpolation which is no longer necessary because of this change.

beccadax committed:
Fix Linux-only build failure

Apparently, the macOS STL is more forgiving about tuples vs. pairs than the Linux one.

beccadax committed:
Benchmark interpolation wtih custom types

beccadax committed:
Add default ExpressibleByStringLiteral.init(stringLiteral:)

Also adds an unavailable version which is applied when the `init(stringInterpolation: DefaultStringInterpolation)` default is used, ensuring that users still need to implement init(stringLiteral:) themselves in that case.

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
[TEMPORARY] Measure init(stringLiteral:) overhead

Temporarily disabling the default implementations for ExpressibleByStringLiteral types using DefaultStringInterpolation so that I can measure the effect of the unconstrained default on the CustomStringNoInterpolation test. Will revert when it’s done.

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Add missing @inlinable annotation

Probably responsible for at least part of performance gap.

brentdax commented:
At least a small portion of the performance gap comes from a missing `@inlinable` annotation, but I'm not sure how much yet.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Correct arguable CustomStringInterpolation benchmark flaw

As currently written, the optimizer can completely remove the involvement of CustomString in some cases. We probably don’t want that, so let’s fix it.

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Test inlining an empty-string check

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark

beccadax committed:
Inline the SmallString check for reserveCapacity()

brentdax commented:
@swift-ci please smoke benchmark

brentdax commented:
@swift-ci please smoke benchmark
